Add Random.full_int

dra27 opened the pull request:
Implements the request in #9487, though using the existing function.

dra27 committed:
Allow 62-bit bound on Random.int on 64-bit systems

xavierleroy commented:
This is misguided.  It is important to me that `Random.int` returns the same results on 32- and 64-bit platforms.

gasche commented:
I am a bit nervous about this:

1. The previous API provided the guarantee that Random.int would behave the same on 32bit and 64bit systems; a working Random.int-using program on either systems would also work on the other and give the same results from the same seed. We should think about whether we want to get rid of this restriction; it may be better to separate the two functions by providing a non-portable `int63` function (or `int60`) as suggested, or by providing an `int64`-returning function.

2. The `int63aux` implementation is relatively inefficient, it appears to consume 90 bits of the random generator. It may be more reasonable to provide a function that only returns 60 bits of entropy, or at least to advertise that this function is slower than calling `int` twice.

3. The 63 naming is a bit awkward. We can only represent positive integers between 0 and 2^62 excluded, so this should probably be int62.

gasche commented:
(It would probably also make sense to consider using a random number generator designed from the start for 63 or 64-bits results, and accept that it is going to be much slower on 32bit systems. Note that this can easily be provided as a third-party library.)

xavierleroy commented:
Concerning more modern PRNGs, and blowing my own horn, there is [PRINGO](https://github.com/xavierleroy/pringo/).

mshinwell commented:
@xavierleroy Interesting, I had never heard of that library :)

gasche commented:
Well if it's not on opam, it does not really exists, so this may be why you never heard of it.

(*goes looking at Xavier's github repositories*)

dra27 commented:
@xavierleroy - I'm puzzled, if not bothered, as to why it's misguided. I've extended the function in a way which fails on 32-bit systems if used in the 64-bit way - any correct program before had to do something to observe the `0x3FFFFFFF` bound on a 64-bit systems and the behaviour of such programs is still identical on both platforms.

dra27 commented:
(by which I mean that for bounds less than `0x3FFFFFFF` the behaviour of 32-bit and 64-bit platforms is identical)

dra27 commented:
@gasche - regarding point 2, please look at the other aux functions, in particular the `int64` generator.

jberdine commented:
Coincidentally, earlier this week I fixed some old code that was helpfully failing on `Int.random max_int`. If I understand correctly, with this patch, such code would not fail, but would silently have slower and different behavior on 64 bit systems. Right? So I would rather see a separate function that can be slower in cases where more bits are needed.

dra27 commented:
@jberdine - slower only inasmuch as in that case you're asking for bigger numbers (the behaviour is the same as for `Random.int64`). I imagine that were one writing the `Random` module today (i.e. both 32 and 64-bit systems), you'd want `Random.int` to be like `Random.{nativeint,int32,int64}` and have a special case `Random.int31`, but I take the example of actually wanting `Random.int` to fail!

dra27 committed:
Add Random.int63

Behaves equivalently on 64-bit and JavaScript and allows 62-bit
upperbound on 64-bit systems.

dra27 commented:
`Random.int63` version pushed. Answering @gasche's other two points: there is already `Random.int64`. For the other functions, the names and auxiliaries reflect the type they return, not the range they accept, so `int63` seems more consistent than `int62`.

gasche commented:
(Minor point: we might have an issue if we later decide to get [Int63](https://ocaml.janestreet.com/ocaml-core/latest/doc/base/Base/Int63/index.html) in the standard library.)

damiendoligez commented on stdlib/random.mli:
```diff
@@ -42,6 +42,12 @@ val int : int -> int
      than 2{^30}. *)
 
+val int63 : int -> int
+(** [Random.int63 bound] is identical to {!int} if [bound] is less than 2{^30},
+    however [bound] may be any positive [int].
```
```suggestion
(** [Random.int63 bound] is equal to [{!int} bound] if [bound] is less than
    2{^30}, however [bound] may be any positive [int].
```

xavierleroy replied:
I would be even more explicit:
```
[Random.int63 bound] returns a random integer between 0 (inclusive)
and [bound] (exclusive). [bound] may be any positive integer. 
If [bound] is less than 2{^30}, [Random.int63 bound] is equal to [{!Random.int} bound].
On a 64-bit platform, [bound] can be greater than 2{^30}, in which case [{!Random.int} bound] would fail but [Random.int63 bound] works. *)
```
The English needs work, but you get the idea: first say what the function does, only then compare and contrast with `Random.int`.

dra27 replied:
How about this?

```diff
--- a/stdlib/random.mli
+++ b/stdlib/random.mli
@@ -43,8 +43,13 @@ val int : int -> int
      than 2{^30}. *)

 val int63 : int -> int
-(** [Random.int63 bound] is identical to {!int} if [bound] is less than 2{^30},
-    however [bound] may be any positive [int].
+(** [Random.int63 bound] returns a random integer between 0 (inclusive)
+     and [bound] (exclusive). [bound] may be any positive integer.
+
+     If [bound] is less than 2{^30}, [Random.int63 bound] is equal to
+     {!Random.int}[ bound]. If [bound] is greater than 2{^30} (on 64-bit systems
+     or non-standard environments, such as JavaScript), [Random.int63] returns
+     a value, where {!Random.int} raises {!Invalid_argument}.
```

I started with this, wrote the JavaScript comment, and then realised that there is a problem on JavaScript for bounds > 0x3FFFFFFF and <= 0x7FFFFFFF

damiendoligez commented on stdlib/random.ml:
```diff
@@ -93,6 +93,22 @@ module State = struct
+  let rec int63aux s n =
+    let b1 = bits s in
+    let b2 = bits s lsl 30 in
+    let b3 = (bits s land 3) lsl 60 in
```
```suggestion
    let b3 = (bits s lsr 28) lsl 60 in
```
It's better to use the high-order bits because (as far as I remember) they are slightly more random.

damiendoligez reviewed (approved):
Two suggestions. I'm approving in principle but please get @xavierleroy's signoff on the new API.

xavierleroy reviewed (commented):
OK, possibly with more work on the documentation comment.

Not quite related, but I often miss `bits32`, `bits64`, `nativebits` functions in Random.

stedolan commented on stdlib/random.ml:
```diff
@@ -93,6 +93,22 @@ module State = struct
+  let rec int63aux s n =
+    let b1 = bits s in
+    let b2 = bits s lsl 30 in
+    let b3 = (bits s land 3) lsl 60 in
```
This does the wrong thing on 32-bit systems. Left shifting by `60` is undefined, but comes out as left shifting by `28` on x86. The top four bits of `r` are then the bitwise OR of two uniform random numbers, so end up with a 25% / 75% probability of being 0 or 1, instead of being uniform.

(Arguably this serves you right for calling int63 on a 32-bit system, but it is still pretty weird.)

stedolan replied:
Actually, never mind, I think you can never hit this path on 32-bit systems? You'd need to construct a bound > 0x3FFFFFFF.

xavierleroy replied:
Yes, exactly, `int63aux` is never called on a 32-bit host.

dra27 replied:
Actually, I was just about to push a correction and ask a question. This does create a problem on JavaScript, and I'm not sure what our opinion on that should be?

xavierleroy replied:
What problem on JavaScript?

dra27 replied:
It hits exactly this problem, for example with a bound of 0x7FFFFFFF (which is valid in js_of_ocaml)

dra27 replied:
Well, technically it's not exactly this problem - it hits something like it because of representing integers as doubles :)

xavierleroy replied:
> representing integers as doubles :)

Terrible idea, indeed :-)  ~~The logical "or" of `b1`, `b2`, `b3` will have some low bits always equal to 0.  I would guess the 9 low bits are 0 with probability 1/2, 8 low bits with 3/4, 7 low bits with 7/8, etc.  (I may be off by one.)~~

Where do we go from here?

xavierleroy replied:
Actually, I don't understand how ocaml-to-javascript translators compile integer arithmetic.  Is it the case that all integer operations are taken modulo 2^32?

stedolan replied:
I don't think that's quite the problem: Javascript's bitwise or happens on 32-bit integers. (It's roughly: convert double to integer, truncate down to 32 bits, bitwise-or, convert back to double. 32-bit ints can roundtrip through double losslessly, so a sequence of bitwise operations gives the same answer as on actual 32-bit numbers).

xavierleroy replied:
OK, that was not the problem.  Is it the left shift by 60, then?  Because this can be avoided by factoring:
```
    let b1 = bits s in
    let b2 = bits s in
    let b3 = bits s land 3 in
    let b = ((b2 lor (b3 lsl 30)) lsl 30) lor b1
```

dra27 replied:
I don't precisely get how it works either! In terms of implementation, is having 0x3FFFFFFF < bound <= 0x7FFFFFFF still get 90 random bits but only mix 31 of them disastrous?!

xavierleroy replied:
As @stedolan wrote, if the left shift by 60 is interpreted as a left shift by 28, you get a bias towards 1 on bits 28, 29, ...  Also, and even with my factoring trick, you can end up with a negative 32-bit integer.

xavierleroy replied:
Another proposal: first we get an int with all bits random, then we right shift by 1 to make it nonnegative.
```
    let b1 = bits s in
    let b2 = bits s in
    let b3 = bits s in
    let b = (((b3 lsl 30) lor b2) lsl 30) lor b1 in (* 63 random bits on OCaml 64 bits, 32 random bits on Javascript *)
    let bpos = b lsr 1 in  (* random nonnegative int *)
    ...

dra27 replied:
Does this work for the entire function, therefore?

```ocaml
  let rec int63aux s n =
    let b1 = bits s in
    let b2 = bits s in
    let b3 = bits s in
    (* 63 random bits on 64-bit OCaml; 32 random bits on JavaScript *)
    let b = ((((b3 land 0x38000000) lsl 3) lor b2) lsl 30) lor b1 in
    let bpos = b lsr 1 in (* Random non-negative int *)
    let r =
      (* If 0x3FFFFFFF < n <= 0x7FFFFFFF, take only the low 31 bits,
         for consistent results between 64-bit OCaml and 32-bit JavaScript. *)
      if n < 1 lsl 31 then
        bpos land ((1 lsl 30) + 0x3FFFFFFF)
      else
        bpos
    in
    let v = r mod n in
    if r - v > max_int - n + 1 then int63aux s n else v
```
It's your second proposal, but using the top 3 bits of `b3` as per @damiendoligez's comment.

xavierleroy replied:
I don't understand the special case `if n < 1 lsl 31`. But if you must special-case `0x3FFFFFFF < n <= 0x7FFFFFFF`, you could just as well combine two `bits s` to get a 31-bit number directly, rather than combine three `bits s` then chop down more bits.

dra27 replied:
Without the special case, you get a different `r mod n` between JavaScript and 64-bit bytecode?

For mixing two calls to `bits` instead of three, that was my previous question - is it better that we _always_ get 90 bits for bounds above `0x3FFFFFFF`, or does it not matter if bounds between `0x3FFFFFFF` and `0x7FFFFFFF` get 60 bits and only bounds above `0x7FFFFFFF` get 90?

alainfrisch replied:
> > representing integers as doubles :)
> 
> Terrible idea, indeed :-)

(Side-note) You probably know that, but there are no "integer" values in Javascript, just "numbers" (=doubles).  The language still "knows" about integers through its bitwise operators `|`, `&`, `^` which interpret their arguments and result as 32-bit integers.  `(x1 + x2)|0` is thus the standard way to emulate a 32-bit integer semantics.

xavierleroy replied:
> For mixing two calls to bits instead of three, that was my previous question - is it better that we always get 90 bits for bounds above 0x3FFFFFFF, or does it not matter if bounds between 0x3FFFFFFF and 0x7FFFFFFF get 60 bits and only bounds above 0x7FFFFFFF get 90?

Sorry if I misunderstood the question.  For bounds up to 0x7FFF_FFFF, 60 bits of randomness are plenty to produce a 31-bit integer that will be then reduced to the 0...bound range.  If the bound is just above 0x4000_0000 you may end up iterating more than if you started with a 60-bit random integer, but still it should converge exponentially quickly.

dra27 commented:
Everything should now be addressed. I couldn't quickly get `js_of_ocaml` working with trunk, but it's easy to extract Random.State.{bits,intaux,int63aux,int63} and run it on 4.11. This program:

```ocaml
let s = Random.State.make [| 42 |] in
Printf.printf "%x\n" (int63 s 0x6FFFFFFF);
```
correctly displays `3dd6aa13` on both 64-bit bytecode and 32-bit JavaScript. If the `if n <= max_int_32 then` special case is removed, 64-bit bytecode gives `1d4172ec` and 32-bit JavaScript gives `29ca4560`

xavierleroy commented on stdlib/random.ml:
```diff
@@ -93,6 +93,31 @@ module State = struct
+    let (b, max_int) =
+      if n <= max_int_32 then
+        (* 32 random bits on both 64-bit OCaml and JavaScript *)
+        (((b2 lsl 30) lor b1) land max_int_32, max_int_32)
```
@damiendoligez prefers the high bits to the low bits, so what about
```suggestion
        ((((b2 land 0x30000000) lsl 2) lor b1) land max_int_32, max_int_32)
```
Pushing his logic further, we should combine the 16 top bits from b1 and from b2, and something similar in the b3 case below, but maybe that's enough for now.

xavierleroy replied:
@dra27 could  you please accept the suggestion or reject it explicitly?  We need to make progress with PRs.

xavierleroy reviewed (approved):
Looks good to me.  One more suggestion below...

dra27 committed:
Use upper bits of each 30-bit random int

dra27 commented:
Sorry for the slowness. I took your suggestion, which nicely means that makes both versions consistent. FWIW, I pushed an extra commit which uses the top 16 bits of each for the `< 0x7FFFFFFF` case and the top 21 bits of each for the 63-bit case.

xavierleroy commented on stdlib/random.ml:
```diff
@@ -93,6 +93,39 @@ module State = struct
+        (* 32 random bits on both 64-bit OCaml and JavaScript.
+           Use upper 16 bits of b1 and b2. *)
+        let bpos =
+          (((b2 land 0x3FFFC000) lsl 2) lor (b1 lsr 14)) land max_int_32
```
I missed the `land max_int_32` in my previous reviews but I'm afraid it's wrong.  The effect is to zero bit number 31, but the `let r = b lsr 1` below will then zero bit number 30, and you end up with a candidate between 0 and 2^30 instead of between 0 and 2^31, and random values above 2^30 will never be generated.

xavierleroy replied:
I took the liberty of pushing to your branch a new test that performs chi-square testing on the functions from module Random.  You'll see that the test `"Random.int63 (256 * p) / p"` fails the test.  It exercises the case 2^30 < n < 2^31.

xavierleroy replied:
I updated the test following @Octachron 's comments (concerning the confidence interval) and to test the js-of-ocaml model (integer size = 32 bits, word size = 32 bits) as well.

xavierleroy reviewed (commented):
Thanks for the update, but I'm afraid something is not right, see below.

There is no way this code can get merged without testing against Dieharder or similar PRNG checker. There are some tests in my PRINGO project; I'll see what can be done along these lines.

xavierleroy committed:
Simple chi-square test for functions from module Random

xavierleroy commented:
This PR is stalled, apparently.  I split the chi^2 state off and merged it in trunk, as this test is useful by itself.

damiendoligez commented on stdlib/random.ml:
```diff
@@ -93,6 +93,39 @@ module State = struct
+        (* 32 random bits on both 64-bit OCaml and JavaScript.
+           Use upper 16 bits of b1 and b2. *)
+        let bpos =
+          (((b2 land 0x3FFFC000) lsl 2) lor (b1 lsr 14)) land max_int_32
```
```suggestion
        (* 31 random bits on both 64-bit OCaml and JavaScript.
           Use upper 15 bits of b1 and 16 bits of b2. *)
        let bpos =
          ((b2 land 0x3FFFC000) lsl 1) lor (b1 lsr 15)
```
Since you need only 31 bits, why not generate 31 bits directly? See below.

damiendoligez commented on stdlib/random.ml:
```diff
@@ -93,6 +93,39 @@ module State = struct
+        (* 63 random bits on 64-bit OCaml; unreachable on JavaScript.
+           Use upper 21 bits of b1, b2 and b3. *)
+        let bpos =
+          ((((b3 land 0x3FFFFE00) lsl 12) lor (b2 lsr 9)) lsl 21) lor (b1 lsr 9)
```
```suggestion
        (* 62 random bits on 64-bit OCaml; unreachable on JavaScript.
           Use upper 21 bits of b1 and b2 and 20 bits of b3. *)
        let bpos =
          ((((b3 land 0x3FFFFE00) lsl 12) lor (b2 lsr 9)) lsl 20) lor (b1 lsr 10)
```
likewise, generate only 62 bits

damiendoligez commented on stdlib/random.ml:
```diff
@@ -93,6 +93,39 @@ module State = struct
+        in
+          (bpos, max_int)
+    in
+    let r = b lsr 1 in (* Random non-negative int *)
```
```suggestion
```
Then you can get rid of this operation.

damiendoligez reviewed (commented):
Suggestion for shortening the code while fixing the bug found by Xavier.

xavierleroy reviewed (dismissed):
Since I cannot figure out how to remove my previous approval, I'm going to request changes. The changes in question were described in details by @damiendoligez and I in our reviews.

dra27 committed:
Fix 32-bit path (unnecessary use of wrong mask)

dra27 committed:
Add a test for the JavaScript case

dra27 committed:
Remove the unnecessary extra bit

dra27 commented:
OK, finally. I've rebased but put everything changed into subsequent commits - I'll squash those when we're done before merging.

The `land max_int_32` should have been `land (1 lor max_int_32 lsl 1)` (i.e. `0xFFFFFFFF`) but in fact it was unnecessary and must been left-over from something previously - the expression was always 32 bits (both on JavaScript and 63-bit OCaml). That's fixed in https://github.com/ocaml/ocaml/pull/9489/commits/9352d81391120517883e7384c1d2604726e0ec2c and `Random.int63` then passes the χ² test. Xavier suggested generating 1 extra bit and right-shifting in https://github.com/ocaml/ocaml/pull/9489#discussion_r491915347 as part of getting an expression which was positive on both 32-bit JavaScript and 63-bit OCaml... but we still ended up with a special case for the `0x3FFFFFFF < bound <= 0x7FFFFFFF` since we need a 32-bit positive number on **both** 32-bit JavaScript and 64-bit OCaml, and so the common expression was removed, but the right shifting remained. So @damiendoligez is (unsurprisingly) correct, and I took the suggestion to eliminate the extra bit in https://github.com/ocaml/ocaml/pull/9489/commits/414ff8c362cca843e90c328081278350bd3922d4 (modulo correcting the comment for the 64-bit case - it's 20 bits of `b1`, rather than `b3`).

I've also added a single extra test in https://github.com/ocaml/ocaml/pull/9489/commits/efd6c4596507f28e7ef83ee893f800306f775c53 which can also be checked using `js_of_ocaml`.

I think this one's finally ready to go... sorry for the delays between rounds of review.

favonia commented:
> (Minor point: we might have an issue if we later decide to get [Int63](https://ocaml.janestreet.com/ocaml-core/latest/doc/base/Base/Int63/index.html) in the standard library.)

My 2¢: how about `Random.full_int` like `Gc.full_major` or some name other than `intNN`? This way, we would not have to worry about the future when 128-bit operations become cheaper. Otherwise, perhaps we should say `Random.Int63` only accepts a `bound` less than 2^62 instead of "any positive integer".

damiendoligez commented:
>how about Random.full_int

I like this suggestion. The discrepancy with `Random.int32` and `Random.int64` is disturbing and indeed the potential conflict with `Int63` is annoying.

dra27 commented:
I don't mind - it'd obviously be good finally to get this PR merged!

> I like this suggestion. The discrepancy with `Random.int32` and `Random.int64` is disturbing and indeed the potential conflict with `Int63` is annoying.

I'm having a slow Monday morning - the conflict?

favonia commented:
> I'm having a slow Monday morning - the conflict?

@gasche pointed out that, if we ever introduce `int63` to OCaml, then we probably want `Random.int63` for `int63`.

I really like `full_int` (or some name without the number `63`) because its semantics is a drop-in replacement of `Random.int` (unless you want to keep the bound).

dra27 committed:
Random.full_int instead of Random.int63

dra27 commented:
> > I'm having a slow Monday morning - the conflict?
> 
> @gasche pointed out that, if we ever introduce `int63` to OCaml, then we probably want `Random.int63` for `int63`.
> 
> I really like `full_int` (or some name without the number `63`) because its semantics is a drop-in replacement of `Random.int` (unless you want to keep the bound).

Thanks! I'm having a slightly faster Tuesday morning 🙂

Change to `Random.full_int` pushed - FWIW I left the internal function called `int63_aux` since that not only describes what it does (it crashes with 31 bit ints) but were we to add int63 as a type this function would probably be simplified to use it (and/or it would be used for the actual `Random.int63`).

damiendoligez commented:
I think all of @xavierleroy's objections have been addressed, so I took the liberty of dismissing his change request. Let's merge now!
